Code/GraphMol/MolPickler.cpp

```diff
@@ -801,55 +802,91 @@ void MolPickler::pickleMol(const ROMol &mol, std::string &ss) {
   pickleMol(&mol, ss, MolPickler::getDefaultPickleProperties());
 }
 
+// Resets the `exceptionState` of the `istream` in the destructor to the
+// `exceptionState` the `istream` was in the constructor.
+struct IoExceptionStateResetter {
+  std::ios_base::iostate originalExceptionState;
+  std::istream &originalStream;
+  IoExceptionStateResetter(std::istream &ss)
+      : originalStream(ss), originalExceptionState(ss.exceptions()) {}
+
+  ~IoExceptionStateResetter() {
+    if (originalStream) {
+      originalStream.exceptions(originalExceptionState);
+    }
+  }
+};
+
 // NOTE: if the mol passed in here already has atoms and bonds, they will
 // be left intact.  The side effect is that ALL atom and bond bookmarks
 // will be blown out by the end of this process.
 void MolPickler::molFromPickle(std::istream &ss, ROMol *mol) {
   PRECONDITION(mol, "empty molecule");
-  int32_t tmpInt;
 
-  mol->clearAllAtomBookmarks();
-  mol->clearAllBondBookmarks();
-
-  streamRead(ss, tmpInt);
-  if (tmpInt != endianId) {
-    throw MolPicklerException(
-        "Bad pickle format: bad endian ID or invalid file format");
-  }
-
-  streamRead(ss, tmpInt);
-  if (static_cast<Tags>(tmpInt) != VERSION) {
-    throw MolPicklerException("Bad pickle format: no version tag");
-  }
-  int32_t majorVersion, minorVersion, patchVersion;
-  streamRead(ss, majorVersion);
-  streamRead(ss, minorVersion);
-  streamRead(ss, patchVersion);
-  if (majorVersion > versionMajor ||
-      (majorVersion == versionMajor && minorVersion > versionMinor)) {
-    BOOST_LOG(rdWarningLog)
-        << "Depickling from a version number (" << majorVersion << "."
-        << minorVersion << ")"
-        << "that is higher than our version (" << versionMajor << "."
-        << versionMinor << ").\nThis probably won't work." << std::endl;
-  }
-  majorVersion = 1000 * majorVersion + minorVersion * 10 + patchVersion;
-  if (majorVersion == 1) {
-    _depickleV1(ss, mol);
-  } else {
-    int32_t numAtoms;
-    streamRead(ss, numAtoms, majorVersion);
-    if (numAtoms > 255) {
-      _depickle<int32_t>(ss, mol, majorVersion, numAtoms);
+  // Ensure that the exception mask of the `istream` is reset to the previous
+  // state after we're done.
+  IoExceptionStateResetter resetter(ss);
+  
+  // Enable exceptions, so we're notified when we've reached EOF or any other
+  // problem.
+  ss.exceptions(std::ifstream::eofbit | std::ifstream::failbit |
```

Comment:
I like this, but I'd rather have the exception state passed into the IoExceptionStateRestter:

```
 IoExceptionStateResetter resetter(ss,
   std::ifstream::eofbit | std::ifstream::failbit | std::ifstream::badbit);
```

Which feels more RAII to me for some reason,